#[derive(Property, Export)] for enums

patataofcourse opened the pull request:
MVP macros for deriving Property and Export in enums.

Not done yet, but hoping for reviews since this is my first contribution to a library of this size.

### Commits will be squashed later, please review as one for now.

Current requirements for deriving these traits:
- Must be an enum
- Must have an explicit `#[repr(i*/u*)]` type
- Variants must be unit/fieldless and have explicit discriminant numbers (`A = 2` instead of `A(u32) = 2` or `A`)

From earlier attempts, it seemed like ToVariant and FromVariant were needed as well? Doesn't seem to be the case anymore, weirdly.

TODO list:

- [x] `#[derive(Property)]`
    - [x] Implement
    - [x] Create tests
    - [x] Document
    - [ ] Expand support?
- [x] `#[derive(Export)]`
    - [x] Implement
    - [x] Create tests
    - [x] Document
    - [ ] Expand support?

GodotRust commented:
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-371

Bromeon commented on itest/rust/src/property_test.rs:
```diff
@@ -289,3 +290,28 @@ struct CheckAllExports {
+#[itest]
+fn derive_property() {
+    let mut class = DeriveProperty { foo: Test::B };
+    assert_eq!(class.get_foo(), Test::B as i64);
+    class.set_foo(Test::C as i64);
+    assert_eq!(class.foo, Test::C);
```
I wonder if it's possible that `get_foo/set_foo` return/accept directly the enum in Rust, not the `i64`?

Although maybe that's more complex; in that case it could be done later as a QoL improvement.

patataofcourse replied:
The scope of this PR is more to add the proc macros than to modify library code - I'm not saying I won't be doing this or that it's completely outside the scope of this PR, but this is definitely not a priority for me rn.

Bromeon reviewed (commented):
Thanks a lot for this! 👍 

Really nice that you even added generic support! Do you think it's possible to list one generic example in the test as well, to make sure the syntax is parsed correctly and works?

patataofcourse commented on godot-macros/src/derive_property.rs:
```diff
@@ -0,0 +1,114 @@
+    let body_set;
+    let mut intermediate = quote! {()};
+
+    match decl {
+        Declaration::Enum(enum_) => {
```
I thought of that, but I also thought it might be better to leave it like this for future struct implementations- I'm definitely not sure though lol

Bromeon replied:
Let's make it readable now -- if in the future anything else is more readable, then we can always refactor. But we can't possibly know the requirements yet.

patataofcourse commented on godot-macros/src/derive_property.rs:
```diff
@@ -0,0 +1,114 @@
+                    .iter()
+                    .find(|c| c.path.len() == 1 && c.path[0].to_string() == "repr")
+                {
+                    Some(c) => c.value.clone().into_token_stream(),
```
c.value is still required since otherwise it would take the entirety of the argument, otherwise resolved

patataofcourse commented:
Aside from the stuff I resolved in the reviews (if it's not resolved feel free to reopen), I removed generic support (since it's not usable in any way right now).
I also changed some other stuff then immediately reverted it because it broke. oops.

Bromeon commented on godot-macros/src/derive_property.rs:
```diff
@@ -0,0 +1,118 @@
+        let mut matches_get = quote! {};
+        let mut matches_set = quote! {};
+        intermediate = if let Some(attr) = enum_.attributes.iter().find(|attr| {
+            attr.path.len() == 1
+                && attr.get_single_path_segment().map(|path| path.to_string())
+                    == Some("repr".to_string())
```
The `attr.path.len() == 1` condition is already checked by `get_single_path_segment()`, so you can omit it.

You can also simplify it a bit:
```suggestion
            attr.get_single_path_segment() == Some(&ident("repr"))
```

Bromeon commented on godot-macros/src/derive_property.rs:
```diff
@@ -0,0 +1,118 @@
+                _ => {
+                    return bail!(
+                        v_name,
+                        "Property can only be derived on enums with fully unit variants for now"
```
```suggestion
                        "Property can only be derived on enums with unit variants for now"
```

patataofcourse replied:
I think I meant to type "only" in there - would that work better?

Bromeon replied:
Yep 👍

patataofcourse commented on godot-macros/src/derive_property.rs:
```diff
@@ -0,0 +1,118 @@
+        Declaration::Struct(s) => {
+            return bail!(s.name, "Property can only be derived on enums for now")
+        }
+        _ => unreachable!(),
```
I see - however, all other derive macros I can see (GodotClass, ToVariant, FromVariant) do this as well, should I open an issue?

patataofcourse replied:
Also: should I mention that union support is less likely than struct support?

Bromeon replied:
> I see - however, all other derive macros I can see (GodotClass, ToVariant, FromVariant) do this as well, should I open an issue?

I'll fix it right away, thanks for letting me know!



> Also: should I mention that union support is less likely than struct support?

No, I think `"Property can only be derived on enums for now"` is good enough. We can always clarify what is supported or not once we flesh it out a bit more 🙂

Bromeon commented on godot-macros/src/lib.rs:
````diff
@@ -455,14 +456,56 @@ pub fn derive_from_variant(input: TokenStream) -> TokenStream {
+///
+/// Example :
+///
+/// ```ignore
````
Could you use _no_run_ instead of _ignore_?

That way, the code is still compiled. Use the `#` placeholder to hide imports like the prelude.

Bromeon commented on godot-macros/src/derive_export.rs:
```diff
@@ -0,0 +1,80 @@
+                );
+            };
+
+            hint_string += &format!(
+                "{v_name}:{}{}",
+                v_disc.to_string().trim_matches(['(', ')'].as_slice()),
+                if i < enum_.variants.inner.len() - 1 {
+                    ","
+                } else {
+                    ""
+                }
+            );
```
Could you extract the format arguments into separate variables? Then you can substitute them inline like in `{vname}`.

Also, maybe [`join()`](https://doc.rust-lang.org/std/primitive.slice.html#method.join) helps with the intermediate comma?

Bromeon commented on godot-macros/src/lib.rs:
```diff
@@ -455,14 +456,56 @@ pub fn derive_from_variant(input: TokenStream) -> TokenStream {
+/// - The enum variants must have explicit discriminants, that is, e.g. `A = 2`, not just `A`
+///     - This will likely stay this way in order to keep proc-macro dependencies lighter.
+///
+/// Example :
```
```suggestion
# Example
```

Bromeon commented on godot-macros/src/lib.rs:
```diff
@@ -455,14 +456,56 @@ pub fn derive_from_variant(input: TokenStream) -> TokenStream {
+///     foo: TestEnum
+/// }
+///
+/// // This would not panic.
+/// let mut class = TestClass {foo: TestEnum::B}.get_foo();
+/// assert!(class.get_foo() == TestEnum::B as i32);
+/// class.set_foo(TestEnum::A);
+/// assert!(class.foo == TestEnum::A as i32);
```
The "not panic" comment can be removed (does not apply here). Also, `assert_eq` would be nice.

I'm fixing the same things in `ToVariant`/`FromVariant` 🙂

Bromeon commented on godot-macros/src/lib.rs:
```diff
@@ -455,14 +456,56 @@ pub fn derive_from_variant(input: TokenStream) -> TokenStream {
+/// - Only works for enums, structs aren't supported by this derive macro at the moment.
+/// - The enum variants must not have any fields - currently only unit variants are supported.
+/// - The enum variants must have explicit discriminants, that is, e.g. `A = 2`, not just `A`
+///     - This will likely stay this way in order to keep proc-macro dependencies lighter.
```
This last sentence (will likely stay that way) is not yet decided, I would avoid too much speculation about the future in documentation. We're still relatively early with gdext, a lot can happen 🙂

Bromeon reviewed (commented):
Some superficial comments about the new changes, will probably need another look tomorrow...

Bromeon commented on itest/rust/src/property_test.rs:
```diff
@@ -293,3 +294,78 @@ struct CheckAllExports {
+}
+
+#[derive(GodotClass)]
+#[class(base = Object)]
```
You can omit this line. Then the base is implicitly `RefCounted`, which means the object is destroyed at the end of scope. This should fix the memory leak in CI.

Below, you would then need to use `RefCountedVirtual`.

Bromeon commented on godot-macros/src/lib.rs:
```diff
@@ -454,6 +456,60 @@ pub fn derive_from_variant(input: TokenStream) -> TokenStream {
+///     foo: TestEnum
+/// }
+///
+/// # #[godot_api]
+/// # impl TestClass {}
```
Could you add a comment:
```suggestion
/// # // TODO remove when https://github.com/godot-rust/gdext/issues/187 is truly addressed
/// # #[godot_api]
/// # impl TestClass {}
```

Otherwise we keep tech-debt around forever  🙂

Bromeon commented on godot-macros/src/lib.rs:
````diff
@@ -454,6 +456,60 @@ pub fn derive_from_variant(input: TokenStream) -> TokenStream {
+
+/// Derive macro for [Export](godot::bind::property::Property) on enums.
+///
+/// Currently has some tight requirements which are expected to be softened as implementation expands:
+/// - Only works for enums, structs aren't supported by this derive macro at the moment.
+/// - The enum variants must not have any fields - currently only unit variants are supported.
+/// - The enum variants must have explicit discriminants, that is, e.g. `A = 2`, not just `A`
+/// ```
````
Could you maybe refer to the `Property` derive macro for the requirements? Then we only need to define them in one place.

Bromeon reviewed (commented):
Thanks a lot! The CI fails due to memory leaks, see first comment for addressing them.

After addressing the remaining comments and green CI, it should be good to go, so you could already squash the commits!

Bromeon commented:
Thank you! 👍
